### ISSUE TITLE
Multiple DDL statements inside of a migration Up or Down function

### ISSUE THREAD

**Issue by abjrcode** [knex/knex#4678]
# Environment

Knex version: 0.95.11
Database + version: MySQL 5.8
OS: Windows 10

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

If there are multiple DDL statements such as creating multiple tables inside of a single migration script's `up()` function and one of them fails then I expect knex to call that migration script's `down()` to revert/rollback any partial updates that might have took place.

2. Error message
There are none.

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real

Let's assume the following migration script:
```
import { Knex } from 'knex'

export async function up(knex: Knex): Promise<void> {
  await knex.schema.createTable('ValidTable', tableBuilder => {
    tableBuilder.increments('Id', { primaryKey: true })
  })

  await knex.schema.createTable('$#InvalidName', tableBuilder => {
    tableBuilder.increments('Id', { primaryKey: true })
  })
}

export async function down(knex: Knex): Promise<void> {
  await knex.schema.dropTableIfExists('ValidTable')
await knex.schema.dropTableIfExists('$#InvalidName')
}
```

The `up()` function will fail on the second `createTable` statement because it uses an invalid identifier but now I am left with a table called `ValidTable` in my system and running the rollback function won't help because this migration script was never recorded because it wasn't successful.

At the moment I am enforcing a rule to never have multiple DDL statements inside the same migration script to avoid this phenomenon but please let me know if I am missing something here. Obviously care should still be taken inside of the `down()` function not to cause issues but Knex cannot guarantee that nor I think it should.

**Comment by KristjanTammekivi** [comment]
MySQL has a bit strange behavior on this, since it [autocommits after certain actions](https://dev.mysql.com/doc/refman/8.0/en/implicit-commit.html). On other databases that's not the case and the DDL executed in up is simply rolled back. Down migrations are only run to rollback (successfully) completed migrations.

In your case I'd recommend adding a try / catch, calling down from there and then throwing the caught error

**Comment by abjrcode** [comment]
Alright if this is something by design then I understand. Indeed there are two possibilities:
* Do one DDL statement per migration file (which is what I am doing now)
* Like you mentioned, catch the exception and deal with it.

It is messy to deal with both ways.

Thanks for your help.

**Comment by KristjanTammekivi** [comment]
Yeah I guess one ddl per file would actually be nicer, but still not perfect.

**Comment by elhigu** [comment]
I think that when `up()` fails, then `down()` should be called. So I would incline on this being a bug. At least I have also thought that it should be done exactly because it allows to write cleanup code easily for partially done migration and wouldn't require 2 different down handlers (one for exception handled reverting and another for down function).

Anyhow if this is decided not to be a bug, then best practises for writing mysql, sqlite and mssql migrations would be good to write. I would do it in a way that I would have a single try - catch block in top level of migration, which checks where error happened and would revert all the changes before that position.

In either way this should be explained in documentation since this is pretty common pitfall for mysql etc. users.

> MySQL has a bit strange behavior on this, since it autocommits after certain actions. On other databases that's not the case and the DDL executed in up is simply rolled back. Down migrations are only run to rollback (successfully) completed migrations.

Nitpick: After DDL implicit commit happens. Autocommit is a term for a bit different thing :)

**Comment by mercmobily** [comment]
This is expected on MySQL and some other dialects. DDL causes an implicit commit and is not transactional, so a failure midway through multiple DDL statements leaves earlier statements applied. Knex only runs `down()` for fully applied migrations, not for partially applied `up()`.

Two practical recommendations:
-  Keep DDL operations in separate migrations to minimise partial‑application fallout.
- If you need multiple DDLs in one migration on MySQL, wrap the body of `up()` in `try`/`catch`, perform explicit cleanup (calling your own `down` logic) on error, and then rethrow. That mimics transactional behaviour for MySQL’s non‑transactional DDL.

I agree that documentation could be clearer here; we should add a note under migrations about implicit commits on MySQL/MariaDB and best practices. @elhigu also suggested `down()` being called on failed `up()`; that would be a behaviour change and needs discussion/consideration.

I am labeling this as a bug and as a documentation issue. It will need more triaging when we come to solve it.
